### PR TITLE
 Decision strategy: incorporate bounded integers

### PR DESCRIPTION
**Pull request by ajreynol**

### CODE REVIEW

File: src/theory/quantifiers/fmf/bounded_integers.cpp [hunk 1]
```diff
@@ -43,125 +48,51 @@ BoundedIntegers::IntRangeModel::IntRangeModel(BoundedIntegers * bi, Node r, cont
...
+BoundedIntegers::BoundedIntegers(context::Context* c, QuantifiersEngine* qe)
+    : QuantifiersModule(qe)
+{
 }
 
-BoundedIntegers::~BoundedIntegers() { 
-  for( std::map< Node, RangeModel * >::iterator it = d_rms.begin(); it != d_rms.end(); ++it ){
+BoundedIntegers::~BoundedIntegers()
+{
+  for (std::map<Node, IntRangeDecisionHeuristic*>::iterator it = d_rms.begin();
```
Comment (HanielB) [inline src/theory/quantifiers/fmf/bounded_integers.cpp]: foreach

File: src/theory/quantifiers/fmf/bounded_integers.cpp [hunk 2]
```diff
@@ -356,29 +287,26 @@ bool BoundedIntegers::needsCheck( Theory::Effort e ) {
...
-
-void BoundedIntegers::addLiteralFromRange( Node lit, Node r ) {
-  d_lit_to_ranges[lit].push_back(r);
-  //check if it is already asserted?
-  if(d_assertions.find(lit)!=d_assertions.end()){
-    d_rms[r]->assertNode( d_assertions[lit] ? lit : lit.negate() );
+  Trace("bint-engine") << "---Bounded Integers---" << std::endl;
+  bool addedLemma = false;
+  // make sure proxies are up-to-date with range
+  for (unsigned i = 0, size = d_ranges.size(); i < size; i++)
```
Comment (HanielB) [inline src/theory/quantifiers/fmf/bounded_integers.cpp]: foreach

File: src/theory/quantifiers/fmf/bounded_integers.h [hunk 3]
```diff
@@ -83,47 +83,59 @@ class BoundedIntegers : public QuantifiersModule
...
-    virtual Node getNextDecisionRequest() = 0;
-    virtual bool proxyCurrentRange() = 0;
-  };
-  class IntRangeModel : public RangeModel {
+   IntRangeDecisionHeuristic(Node r,
+                             context::Context* c,
+                             context::Context* u,
+                             Valuation valuation,
+                             bool isProxy);
+   virtual ~IntRangeDecisionHeuristic() {}
```
Comment (HanielB) [inline src/theory/quantifiers/fmf/bounded_integers.h]: kill `virtual`

File: src/theory/quantifiers/fmf/bounded_integers.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/quantifiers/fmf/bounded_integers.cpp]: Will clean this up to use unique_ptr instead.

**Pull request merged**